src/box/box.cc

```diff
@@ -4668,6 +4668,10 @@ box_process_subscribe(struct iostream *io, const struct xrow_header *header)
 		if (replica->id != REPLICA_ID_NIL)
 			tnt_raise(ClientError, ER_PROTOCOL, "Can't subscribe "
 				  "an anonymous replica having an ID assigned");
+		if (!replica->anon)
+			tnt_raise(ClientError, ER_PROTOCOL, "Can't subscribe "
+				  "a previously deleted non-anonymous replica "
+				  "as an anonymous replica");
```

Comment:
Are you sure this check is needed?

Why can't we allow to subscribe the replica as anonymous?

Besides, you need another check for `!req.is_anon`: I just managed to:
1. delete replica from _cluster
2. issue `box.cfg{replication=""}` and `box.cfg{replication="old_replication"}` on it
3. and insert it into the just-deleted slot.

Is this expected?

I think that the replica should be able to subscribe anonymously after the _cluster deletion.

cc @Gerold103, @Serpentian

Reply:
We cannot subscribe the deleted replica as anonymous, since the master node sees it as non-anonymous in his replicaset table, and AFAIC, transitioning it to anonymous (instead of deleting it from the replicaset table and inserting it as a new anonymous replica) is not easy, since it will break a lot of invariants.

Reply:
> Is this expected?

Also seems valid to me.

> since it will break a lot of invariants.

The question here why it is forbidden to turn a normal replica into an anonymous one. Maybe @sergepetrenko knows the motivation. Moreover, you won't catch the situation of resubscription if replica object is deleted.

Reply:
Sorry, I didn't explain my point good enough.

Basically what happened on steps 2 and 3, the deleted replica registered in the cluster once again.

Anyway, I thought about this some more and I don't see how we can forbid this.

This only happens if the user restarts replication manually, so this seems fine, indeed.

Reply:
> The question here why it is forbidden to turn a normal replica into an anonymous one. Maybe @sergepetrenko knows the motivation. Moreover, you won't catch this situation of resubscription if replica object is deleted.

This simply was never needed.

Reply:
After oral discussion: this check is legit.